File: SharedSources/SupertonicDaemon.swift
import Foundation

/// Supertonic TTS 데몬 — Python 상주 프로세스와 stdin/stdout 바이너리 프로토콜로 통신
///
/// 프로토콜:
///   Request:  [4 bytes: text_length LE uint32][UTF-8 text]
///   Response: [4 bytes: pcm_length LE uint32][16-bit PCM @ 44100Hz mono]
///
/// 특수 명령: PING, QUIT, VOICE:X, LANG:X, SPEED:X
@available(macOS 14.0, *)
public class SupertonicDaemon: TTSAudioProvider {
    public let sampleRate: Double = 44100
    
    private var process: Process?
    private var stdinHandle: FileHandle?
    private var stdoutHandle: FileHandle?
    private var stderrHandle: FileHandle?
    
    private let pythonPath: String
    private let scriptPath: String
    private var voiceName: String
    private var lang: String
    private var speed: Double
    private var totalSteps: Int
    
    /// 데몬 프로세스 실행 중 여부
    public private(set) var isRunning: Bool = false
    
    /// 동시 접근 방지용 직렬 큐
    private let serialQueue = DispatchQueue(label: "supertonic.daemon.serial")
    
    /// 동시 합성 요청 직렬화용 actor
    private let synthesisLock = SynthesisLock()
    
    public init(
        pythonPath: String? = nil,
        scriptPath: String? = nil,
        voiceName: String = "M1",
        lang: String = "ko",
        speed: Double = 1.05,
        totalSteps: Int = 5
    ) {
        // venv python 경로 자동 탐색
        let bundlePath = Bundle.main.bundlePath
        let projectDir = URL(fileURLWithPath: bundlePath).deletingLastPathComponent().path
        
        if let path = pythonPath {
            self.pythonPath = path
        } else {
            // 프로젝트 내 .venv 탐색
            let venvPython = projectDir + "/.venv/bin/python3"
            if FileManager.default.fileExists(atPath: venvPython) {
                self.pythonPath = venvPython
            } else {
                // 하드코딩된 폴백 경로
                let fallbackVenv = NSString("~/DATA/personal/50_hobbies/super-voice-assistant/.venv/bin/python3").expandingTildeInPath
                if FileManager.default.fileExists(atPath: fallbackVenv) {
                    self.pythonPath = fallbackVenv
                } else {
                    self.pythonPath = "/usr/bin/python3"
                }
            }
        }
        
        if let path = scriptPath {
            self.scriptPath = path
        } else {
            let projectScript = projectDir + "/scripts/supertonic_daemon.py"
            if FileManager.default.fileExists(atPath: projectScript) {
                self.scriptPath = projectScript
            } else {
                let fallbackScript = NSString("~/DATA/personal/50_hobbies/super-voice-assistant/scripts/supertonic_daemon.py").expandingTildeInPath
                self.scriptPath = fallbackScript
            }
        }
        
        self.voiceName = voiceName
        self.lang = lang
        self.speed = speed
        self.totalSteps = totalSteps
    }
    
    deinit {
        stop()
    }
    
    // MARK: - Lifecycle
    
    /// 데몬 프로세스 시작. READY 신호까지 대기.
    public func start() async throws {
        if isRunning { return }
        
        guard FileManager.default.fileExists(atPath: scriptPath) else {
            throw SupertonicError.scriptNotFound(scriptPath)
        }
        
        let proc = Process()
        proc.executableURL = URL(fileURLWithPath: pythonPath)
        proc.arguments = [scriptPath, voiceName, lang, String(speed), String(totalSteps)]
        
        // 환경변수 전달 (venv에서 필요할 수 있음)
        var env = ProcessInfo.processInfo.environment
        env["PYTHONUNBUFFERED"] = "1"
        proc.environment = env
        
        let stdinPipe = Pipe()
        let stdoutPipe = Pipe()
        let stderrPipe = Pipe()
        
        proc.standardInput = stdinPipe
        proc.standardOutput = stdoutPipe
        proc.standardError = stderrPipe
        
        self.stdinHandle = stdinPipe.fileHandleForWriting
        self.stdoutHandle = stdoutPipe.fileHandleForReading
        self.stderrHandle = stderrPipe.fileHandleForReading
        self.process = proc
        
        // 종료 감지
        proc.terminationHandler = { [weak self] process in
            print("⚠️ Supertonic daemon terminated with code \(process.terminationStatus)")
            self?.isRunning = false
        }
        
        try proc.run()
        
        // READY 대기 (최대 30초)
        let ready = try await waitForReady(timeout: 30)
        guard ready else {
            stop()
            throw SupertonicError.startupTimeout
        }
        
        isRunning = true
        print("✅ Supertonic daemon started (python: \(pythonPath), voice: \(voiceName), lang: \(lang))")
    }
    
    /// READY 시그널 대기
    private func waitForReady(timeout: TimeInterval) async throws -> Bool {
        guard let stderr = stderrHandle else { return false }
        
        return try await withCheckedThrowingContinuation { continuation in
            var buffer = Data()
            var resumed = false
            
            // 타임아웃 타이머
            let timer = DispatchSource.makeTimerSource(queue: serialQueue)
            timer.schedule(deadline: .now() + timeout)
            timer.setEventHandler {
                guard !resumed else { return }
                resumed = true
                timer.cancel()
                continuation.resume(returning: false)
            }
            timer.resume()
            
            // stderr 비동기 읽기
            DispatchQueue.global(qos: .userInitiated).async {
                while !resumed {
                    let chunk = stderr.availableData
                    if chunk.isEmpty { break }
                    
                    buffer.append(chunk)
                    if let text = String(data: buffer, encoding: .utf8) {
                        // stderr 로그 출력
                        for line in text.components(separatedBy: "\n") where !line.isEmpty {
                            print("🐍 [supertonic] \(line)")
                        }
                        
                        if text.contains("READY") {
                            guard !resumed else { return }
                            resumed = true
                            timer.cancel()
                            continuation.resume(returning: true)
                            return
                        }
                    }
                }
            }
        }
    }
    
    /// 데몬 종료
    public func stop() {
        guard let proc = process, proc.isRunning else {
            isRunning = false
            return
        }
        
        // QUIT 명령 전송 시도
        if let stdin = stdinHandle {
            let quitBytes = "QUIT".data(using: .utf8)!
            var len = UInt32(quitBytes.count).littleEndian
            let header = Data(bytes: &len, count: 4)
            try? stdin.write(contentsOf: header)
            try? stdin.write(contentsOf: quitBytes)
        }
        
        // 1초 대기 후 강제 종료
        DispatchQueue.global().asyncAfter(deadline: .now() + 1.0) { [weak proc] in
            if let proc = proc, proc.isRunning {
                proc.terminate()
            }
        }
        
        isRunning = false
        process = nil
        stdinHandle = nil
        stdoutHandle = nil
        stderrHandle = nil
        
        print("🛑 Supertonic daemon stopped")
    }
    
    // MARK: - TTSAudioProvider
    
    /// 텍스트를 PCM 오디오 청크 스트림으로 변환
    /// 문장 단위로 쪼개서 각각 합성 → 즉시 yield
    public func collectAudioChunks(from text: String) -> AsyncThrowingStream<Data, Error> {
        AsyncThrowingStream { continuation in
            Task {
                do {
                    let sentences = SmartSentenceSplitter.splitIntoSentences(text)
                    print("📖 [Supertonic] Split into \(sentences.count) sentences")
                    
                    for (index, sentence) in sentences.enumerated() {
                        try Task.checkCancellation()
                        
                        let pcmData = try await self.synthesize(sentence)
                        
                        if !pcmData.isEmpty {
                            print("🎵 [Supertonic] Sentence \(index+1)/\(sentences.count): \(pcmData.count) bytes")
                            continuation.yield(pcmData)
                        }
                        
                        // 문장 간 무음 (0.25초 @ 44100Hz, 16-bit)
                        if index < sentences.count - 1 {
                            let silenceSamples = Int(self.sampleRate * 0.25)
                            let silenceData = Data(count: silenceSamples * 2)
                            continuation.yield(silenceData)
                        }
                    }
                    
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }
    
    // MARK: - Synthesis
    
    /// 단일 텍스트 합성 (직렬화됨)
    public func synthesize(_ text: String) async throws -> Data {
        return try await synthesisLock.run {
            try await self._synthesize(text)
        }
    }
    
    /// 실제 합성 (내부용, synthesisLock 내에서만 호출)
    private func _synthesize(_ text: String) async throws -> Data {
        guard isRunning, let stdin = stdinHandle, let stdout = stdoutHandle else {
            throw SupertonicError.daemonNotRunning
        }
        
        let textData = text.data(using: .utf8)!
        
        // 요청 전송: [4 bytes length][text]
        var len = UInt32(textData.count).littleEndian
        let header = Data(bytes: &len, count: 4)
        
        try stdin.write(contentsOf: header)
        try stdin.write(contentsOf: textData)
        
        // 응답 읽기: [4 bytes pcm_length][pcm_data]
        let respHeader = stdout.readData(ofLength: 4)
        guard respHeader.count == 4 else {
            throw SupertonicError.protocolError("Failed to read response header")
        }
        
        let pcmLen = respHeader.withUnsafeBytes { $0.load(as: UInt32.self).littleEndian }
        
        if pcmLen == 0 {
            return Data()
        }
        
        // PCM 데이터 읽기 (큰 데이터는 여러 번에 나눠 올 수 있음)
        var pcmData = Data()
        pcmData.reserveCapacity(Int(pcmLen))
        
        while pcmData.count < pcmLen {
            let remaining = Int(pcmLen) - pcmData.count
            let chunk = stdout.readData(ofLength: remaining)
            if chunk.isEmpty {
                throw SupertonicError.protocolError("Unexpected EOF reading PCM data")
            }
            pcmData.append(chunk)
        }
        
        // stderr 로그 소비 (블로킹 방지)
        drainStderr()
        
        return pcmData
    }
    
    // MARK: - Commands
    
    /// 음성 스타일 변경
    public func setVoice(_ name: String) async throws {
        try await sendCommand("VOICE:\(name)")
        voiceName = name
    }
    
    /// 언어 변경
    public func setLang(_ newLang: String) async throws {
        try await sendCommand("LANG:\(newLang)")
        lang = newLang
    }
    
    /// 속도 변경
    public func setSpeed(_ newSpeed: Double) async throws {
        try await sendCommand("SPEED:\(newSpeed)")
        speed = newSpeed
    }
    
    /// 특수 명령 전송 (ACK 대기)
    private func sendCommand(_ command: String) async throws {
        guard isRunning, let stdin = stdinHandle, let stdout = stdoutHandle else {
            throw SupertonicError.daemonNotRunning
        }
        
        let cmdData = command.data(using: .utf8)!
        var len = UInt32(cmdData.count).littleEndian
        let header = Data(bytes: &len, count: 4)
        
        try stdin.write(contentsOf: header)
        try stdin.write(contentsOf: cmdData)
        
        // ACK (빈 응답) 읽기
        let ack = stdout.readData(ofLength: 4)
        guard ack.count == 4 else {
            throw SupertonicError.protocolError("Failed to read command ACK")
        }
        
        drainStderr()
    }
    
    // MARK: - Helpers
    
    /// stderr 비동기 drain (로그 출력, 블로킹 방지)
    private func drainStderr() {
        guard let stderr = stderrHandle else { return }
        
        DispatchQueue.global(qos: .utility).async {
            let data = stderr.availableData
            if !data.isEmpty, let text = String(data: data, encoding: .utf8) {
                for line in text.components(separatedBy: "\n") where !line.isEmpty {
                    print("🐍 [supertonic] \(line)")
                }
            }
        }
    }
    
    /// 현재 설정 정보
    public var info: String {
        "SupertonicDaemon(voice=\(voiceName), lang=\(lang), speed=\(speed), running=\(isRunning))"
    }
}

// MARK: - SynthesisLock (동시 요청 직렬화)

private actor SynthesisLock {
    func run<T>(_ body: () async throws -> T) async throws -> T {
        return try await body()
    }
}

// MARK: - Errors

public enum SupertonicError: Error, LocalizedError {
    case scriptNotFound(String)
    case startupTimeout
    case daemonNotRunning
    case protocolError(String)
    case synthesisError(String)
    
    public var errorDescription: String? {
        switch self {
        case .scriptNotFound(let path):
            return "Supertonic daemon script not found: \(path)"
        case .startupTimeout:
            return "Supertonic daemon failed to start within timeout"
        case .daemonNotRunning:
            return "Supertonic daemon is not running"
        case .protocolError(let msg):
            return "Supertonic protocol error: \(msg)"
        case .synthesisError(let msg):
            return "Supertonic synthesis error: \(msg)"
        }
    }
}

File: scripts/supertonic_daemon.py
#!/usr/bin/env python3
"""
Supertonic TTS Daemon
=====================
Swift 앱에서 stdin/stdout 바이너리 프로토콜로 통신하는 상주 프로세스.

프로토콜:
  Request:  [4 bytes: text_length (little-endian uint32)][text_length bytes: UTF-8 text]
  Response: [4 bytes: pcm_length (little-endian uint32)][pcm_length bytes: 16-bit PCM @ 44100Hz mono]

특수 명령:
  "PING"     → "PONG\n" (텍스트 응답, health check)
  "QUIT"     → 프로세스 종료
  "VOICE:X"  → 음성 스타일 변경 (예: "VOICE:F1")
  "LANG:X"   → 언어 변경 (예: "LANG:en")
  "SPEED:X"  → 속도 변경 (예: "SPEED:1.2")

시작 시 stderr로 "READY\n" 출력 (모델 로딩 완료 신호).
"""

import sys
import struct
import time
import numpy as np

def main():
    voice_name = sys.argv[1] if len(sys.argv) > 1 else "M1"
    lang = sys.argv[2] if len(sys.argv) > 2 else "ko"
    speed = float(sys.argv[3]) if len(sys.argv) > 3 else 1.05
    total_steps = int(sys.argv[4]) if len(sys.argv) > 4 else 5

    # 모델 로딩 (1회)
    sys.stderr.write(f"LOADING model... voice={voice_name} lang={lang} speed={speed}\n")
    sys.stderr.flush()

    from supertonic import TTS

    tts = TTS(auto_download=True)
    style = tts.get_voice_style(voice_name=voice_name)
    sample_rate = tts.sample_rate  # 44100

    sys.stderr.write(f"READY sample_rate={sample_rate}\n")
    sys.stderr.flush()

    while True:
        try:
            # 요청 헤더 읽기 (4 bytes)
            header = sys.stdin.buffer.read(4)
            if not header or len(header) < 4:
                break

            text_len = struct.unpack('<I', header)[0]
            if text_len == 0:
                continue
            if text_len > 100_000:  # 안전 제한: 100KB
                sys.stderr.write(f"ERROR: text too long ({text_len} bytes)\n")
                sys.stderr.flush()
                # 빈 응답
                sys.stdout.buffer.write(struct.pack('<I', 0))
                sys.stdout.buffer.flush()
                continue

            text = sys.stdin.buffer.read(text_len).decode('utf-8')

            # 특수 명령 처리
            if text == "PING":
                sys.stdout.buffer.write(b"PONG\n")
                sys.stdout.buffer.flush()
                continue
            elif text == "QUIT":
                sys.stderr.write("QUIT received, exiting.\n")
                sys.stderr.flush()
                break
            elif text.startswith("VOICE:"):
                new_voice = text[6:].strip()
                try:
                    style = tts.get_voice_style(voice_name=new_voice)
                    voice_name = new_voice
                    sys.stderr.write(f"VOICE changed to {new_voice}\n")
                except Exception as e:
                    sys.stderr.write(f"ERROR changing voice: {e}\n")
                sys.stderr.flush()
                # 빈 응답으로 ACK
                sys.stdout.buffer.write(struct.pack('<I', 0))
                sys.stdout.buffer.flush()
                continue
            elif text.startswith("LANG:"):
                lang = text[5:].strip()
                sys.stderr.write(f"LANG changed to {lang}\n")
                sys.stderr.flush()
                sys.stdout.buffer.write(struct.pack('<I', 0))
                sys.stdout.buffer.flush()
                continue
            elif text.startswith("SPEED:"):
                try:
                    speed = float(text[6:].strip())
                    sys.stderr.write(f"SPEED changed to {speed}\n")
                except ValueError:
                    sys.stderr.write(f"ERROR: invalid speed value\n")
                sys.stderr.flush()
                sys.stdout.buffer.write(struct.pack('<I', 0))
                sys.stdout.buffer.flush()
                continue

            # TTS 합성
            t0 = time.time()
            wav, duration = tts.synthesize(
                text,
                voice_style=style,
                lang=lang,
                total_steps=total_steps,
                speed=speed
            )
            elapsed = time.time() - t0

            # float32 → 16-bit PCM
            samples = wav.flatten()
            pcm16 = (samples * 32767).astype(np.int16)
            pcm_bytes = pcm16.tobytes()

            sys.stderr.write(
                f"SYNTH len={len(text)} audio={duration[0]:.2f}s "
                f"pcm={len(pcm_bytes)} elapsed={elapsed:.3f}s\n"
            )
            sys.stderr.flush()

            # 응답: [pcm_length][pcm_data]
            sys.stdout.buffer.write(struct.pack('<I', len(pcm_bytes)))
            sys.stdout.buffer.write(pcm_bytes)
            sys.stdout.buffer.flush()

        except KeyboardInterrupt:
            break
        except Exception as e:
            sys.stderr.write(f"ERROR: {e}\n")
            sys.stderr.flush()
            # 에러 시 빈 응답
            try:
                sys.stdout.buffer.write(struct.pack('<I', 0))
                sys.stdout.buffer.flush()
            except:
                break

    sys.stderr.write("Daemon stopped.\n")
    sys.stderr.flush()


if __name__ == "__main__":
    main()
